Finished registration and checkin commands

paskausks opened the pull request:
Completed `Event` methods for registration and checkin. User isn't allowed to register twice, while duplicate checkins aren't checked.
Now, before linking a steamid64 to a users discord id, it's first checked for existence in the rocketleague.lv database. If it doesn't exist, user is asked to ask an admin to be added before being able to link.
Now, when the *unlink* command is used, the player record is kept, only the `steamid64` field is cleared.

Should event be checked if it's visible, before allowing user to register or check in?
Should registration or checkin times be added?
Should max MMR be rechecked upon checkin? (It's updated upon registration, kinda makes it redundant to keep it in the `players` collection now though)
Should existence in rocketleague.lv database be checked upon registration/checkin?

paskausks committed:
Now when steam id is unlinked, instead of delewting the player record, only the steamid64 is deleted.

paskausks committed:
Added steamid64 validation via rl.lv API before linking

paskausks committed:
Finished registration command and event method

paskausks committed:
Added checkin Event method and chat command

ivoaz commented on commands/event/link.js:
```diff
@@ -30,17 +31,43 @@ module.exports = class LinkCommand extends Command {
+            rllvPlayer = await rllv.RocketLeagueAPI.getPlayer(steamid64);
+        } catch (e) {
+            if (e instanceof rllv.PlayerNotFoundException) {
+                player.unsetSteamId64();
```
Looks like steamid64 is always undefined in this case - no need to unset it.

ivoaz commented on lib/player.js:
```diff
@@ -30,12 +30,12 @@ class Player {
     async getMaxMMR() {
```
I think the method should return the value or be named differently, e.g., updateMaxMMR.

ivoaz commented on lib/event.js:
```diff
@@ -93,12 +93,33 @@ class Event {
     async register(player) {
+        player.checkedIn = false;
 
+        this.players.push(player);
```
I think it would be better to copy necessary values to new object and then push, because the player object is retrieved from the database and may contain unnecessary information.

Also it's good practice to copy/clone objects from arguments to avoid unnecessary mutations (like `checkedIn` property in this case) of the original object.

ivoaz commented on lib/event.js:
```diff
@@ -108,6 +129,14 @@ class Event {
+
+    async playerIsRegistered(discordid) {
+        if (isNaN(discordid)) return null;
+        return await mongodb.collection('events').findOne({
```
Should return boolean instead of player object.

ivoaz commented on lib/player.js:
```diff
@@ -45,6 +45,21 @@ class Player {
 
+    async unsetSteamId64() {
+        this.steamid64 = '';
+        await mongodb.collection('players').updateOne({_id: this._id}, {$set: {
```
Can use `$unset` instead of `$set` to remove the field, otherwise it will still store empty string.

paskausks committed:
Made changes requested in the PR review

unsetSteamId64 no longer called for player which doesn't exist in the
rl.lv database.
Upon registration a new object is pushed into the players array in event
instead of passing the current one.
playerIsRegistered now returns a boolean instead of the fetched event
object.
Renamed getMaxMMR to updateMaxMMR in player.js.
unsetSteamId64 now deletes the key from the player object and uses
$unset instead of setting it to an empty string.

paskausks committed:
Now players shouldn't be able to register or checkin to events which are
invisible.

Also, removed default from link command steamid64 argument.
Removed default checkIn property from Player class constructor.

paskausks committed:
Ensure MaxMMR is stored as an integer

paskausks committed:
Check if steamid64 isn't already taken upon linking accts

paskausks committed:
Added command to view registered players, sorted by descending MMR

paskausks committed:
Added checked in player list.

Also, fixed bug where findCurrentEvent didn't return Event object.
Fixed bug where player.steamId64Taken would always return true.
Refactored embed table generation into separate class.

paskausks committed:
Now player's max MMR will be updated on checkin

paskausks committed:
Now player gets mentioned upon successful registration or checkin

paskausks committed:
Added testing framework

Added mocha to .eslintrc environments.
Fixed eslint messages in project.
Added mocha and rewire to dev-dependencies.
Added npm scripts for linting and testing.
Added tests for Event.create() and stub tests for the rest of Event
methods.

paskausks committed:
Added non-player related method tests for event.js

Added tests for lib/event which aren't player related.
Ensured that findCurrentEvent() will always return the latest event.
Fixed where findVisible() returned cursor instead of array with Event
objects.

paskausks committed:
Finished tests for lib/event.

Added remaining tests for lib/event.
Changed all equality assertions to strict instead of absctract.
Fixed where test data wasn't deleted if any of the assertions fail.
Fixed where hasRegistered() didn't work at all as intended.
Ensured getRegisteredPlayers() and getCheckedInPlayers() returns an
array of Player instances instead of plain objects.

paskausks committed:
Added boilerplate for Player and RocketLeagueAPI tests.

paskausks committed:
Refactored lib/event variable name in event tests.

paskausks committed:
Added Rocket League API tests.

Added tests for lib/rrllv-api.
Added .tern-project to .gitignore.

paskausks committed:
Added tests for Player and RocketLeagueAPI classes.

Fixed where Player.setParams() only updated the database record and not
the instance itself.
Added tests for lib/player.
Added tests for lib/rllv-api.
Added mock module test/lib/mock and RocketLeagueAPI mock class.

paskausks committed:
Added a simple test for TournamentListEmbed.

paskausks committed:
Fixed where connection was lost before test suite could complete.

Removed after-test hooks from event and player tests.
Added global test hook which closes database connection after tests have
run.
This creates a bug though - if running individual tests, which don't
require a database connection, the connection, as well as the process
isn't terminated.

paskausks committed:
Added command test framework.

Added mock classes for Discord.Client, Discord.Collection and CommandoMessage.
Added test for "event-create" command.

paskausks committed:
Added tests for the "event-show" command.

paskausks committed:
Added upcoming event list command code.

paskausks committed:
Made "name" argument mandatory in OpenRegistrationCommand.

Name is no longer an optional argument for "!event-registration-open".
This is because when name isn't provided, Event.findCurrentEvent() is
utilised, which looks for events which are either open for registration
or checkin, but an event which is about to be opened for registration
is neither open for registration or checkin, so Event.findCurrentEvent()
will never return an event which isn't opened for registration yet.

paskausks committed:
Added tests for non-player related event commands.

Added non-player related tests.
Changed wording in registration list command.
Updated Message in mock.js to support testing messages containing
embeds.

paskausks committed:
Added tests for "!link-steam" command.

Added tests for "!link-steam" command and removed trailing spaces.
Fixed where promise didn't resolve when needed to create a new player
record when linking steamid64.
Added constants module for sharing across tests and mock classes.
Added discord id, discord user name and server nickname to mock Message.

paskausks committed:
Finished command tests.

Fixed where promises weren't resolved when updating player mmr in
checkin and register commands.
Added yet unfinished tests for unlink-steam, event-register and
event-checkin.
